src/server/handlers/launchpad.js

```diff
@@ -0,0 +1,236 @@
+  const parsed = parseGitHubUrl(repositoryUrl);
+  if (parsed === null || parsed.owner === null || parsed.name === null) {
+    logger.info(`Cannot parse "${repositoryUrl}"`);
+    return res.status(400).send(RESPONSE_GITHUB_BAD_URL);
+  }
+
+  const uri = `/repos/${parsed.owner}/${parsed.name}/contents/snapcraft.yaml`;
+  const options = {
+    headers: {
+      'Authorization': `token ${req.session.token}`,
+      'Accept': 'application/vnd.github.v3.raw'
+    }
+  };
+  logger.info(`Fetching snapcraft.yaml from ${repositoryUrl}`);
+  return requestGitHub.get(uri, options, (err, response, body) => {
+    if (response.statusCode !== 200) {
+      try {
+        body = JSON.parse(body);
+      } catch (e) {
+        logger.info('Invalid JSON received', err, body);
+        return res.status(500).send(RESPONSE_GITHUB_OTHER);
+      }
+      switch (body.message) {
+        case 'Not Found':
+          // snapcraft.yaml not found
+          return res.status(404).send(RESPONSE_GITHUB_NOT_FOUND);
+        case 'Bad credentials':
+          // Authentication failed
+          return res.status(401).send(RESPONSE_GITHUB_AUTHENTICATION_FAILED);
+        default:
+          // Something else
+          logger.info('GitHub API error:', err, body);
+          return res.status(500).send(RESPONSE_GITHUB_OTHER);
+      }
+    }
+
+    let snapcraftYaml;
+    try {
+      snapcraftYaml = yaml.safeLoad(body);
+    } catch (e) {
+      return res.status(400).send(RESPONSE_SNAPCRAFT_YAML_PARSE_FAILED);
+    }
+    return callback(snapcraftYaml);
+  });
+};
+
+// XXX cjwatson 2016-12-13: We should set an appropriate set of
+// architectures.
+export const newSnap = (req, res) => {
+  getSnapcraftYaml(req, res, snapcraftYaml => {
+    const repositoryUrl = req.body.repository_url;
+    if (!('name' in snapcraftYaml)) {
+      return res.status(400).send(RESPONSE_SNAPCRAFT_YAML_NO_NAME);
+    }
+    const lp_client = getLaunchpad();
+    const username = conf.get('LP_API_USERNAME');
+    logger.info(`Creating new snap for ${repositoryUrl}`);
+    lp_client.named_post('/+snaps', 'new', {
+      parameters: {
+        owner: `/~${username}`,
+        distro_series: `/${DISTRIBUTION}/${DISTRO_SERIES}`,
+        name: `${makeSnapName(repositoryUrl)}-${DISTRO_SERIES}`,
+        git_repository_url: repositoryUrl,
+        git_path: 'refs/heads/master',
+        auto_build: true,
+        auto_build_archive: `/${DISTRIBUTION}/+archive/primary`,
+        auto_build_pocket: 'Updates',
+        store_upload: true,
+        store_series: `/+snappy-series/${STORE_SERIES}`,
+        store_name: snapcraftYaml.name
+      }
+    }).then(result => {
+      logger.info(`Authorizing ${result.self_link}`);
+      lp_client.named_post(result.self_link, 'beginAuthorization', {
+        parameters: { success_url: req.body.success_url }
+      }).then(authUrl => {
+        logger.info(`Authorized ${result.self_link}`);
+        return res.redirect(authUrl);
```

Comment:
`/+snaps?ws.op=new` must send the user through to SSO to authorise the upload of snap builds to the store on their behalf.  This process is necessarily initiated by Launchpad (because it needs to get a macaroon from SCA first, behind the scenes), but we ultimately want to end up back in BSI, hence the new `snap.beginAuthorization` endpoint.  Yes, `success_url` would therefore be something in BSI.

We don't need to store the token on our side; that's done in Launchpad before redirecting to `success_url`.